File: sorting/QuickSort.java
package sorting;

public class QuickSort {
    public static void sort(int[] input) {
        if (input == null)
            throw new IllegalArgumentException("input is null");
        
        if (input.length < 2)
            return;
        
        quickSort(input, 0, input.length-1);
    }
    
    private static void quickSort(int[] input, int start, int end) {
        if (start >= end)
            return;
        
        int pivot = quickSelect(input, start, end);
        quickSort(input, start, pivot-1);
        quickSort(input, pivot+1, end);
    }
    
    private static int quickSelect(int[] input, int start, int end) {
        int key = input[end];
        int lastSmall = start-1;
        
        for (int i = start; i<end; i++) {
            if (input[i] <= key) {
                lastSmall++;
                int temp = input[lastSmall];
                input[lastSmall] = input[i];
                input[i] = temp;
            }
        }
        
        input[end] = input[lastSmall+1];
        input[lastSmall+1] = key;
        
        return lastSmall+1;
    }
}

File: OldOldOld/Sorting/Quicksort.c
#include<stdio.h>

void swap(int* input, int posA, int posB) {
	int temp = input[posA];
	input[posA] = input[posB];
	input[posB] = temp;
}

void quickSort(int* input, int start, int end) {
	if (end-start <=0) 
		return;
	
	int key = end;
	int smallest = start-1;
	int runningPointer = start;

	while(runningPointer < key) {

		if (input[runningPointer] <= input[key]) {
			swap(input, smallest+1, runningPointer);
			smallest++;
			runningPointer++;
		}

		else {
			runningPointer++;
		}
	}

	swap(input, smallest+1, key);

	quickSort(input, start, smallest);
	quickSort(input, smallest+2, end);
}

void printArray(int* input, int len) {
	int i;
	for (i=0; i<len; i++)
		printf("%d ", input[i]);
}

int main() {
	//int input[] = { 9, 1, 4, 2, 7, 10, 13, 3, 6};
	//quickSort(input, 0, 8);
	//printArray(input, 9);

	//int input[] = { 9, 1};
	//quickSort(input, 0, 1);
	//printArray(input, 2);

	int input[] = {9};
	quickSort(input, 0, 0);
	printArray(input, 1);

	return 0;
}
